x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/queries/SemanticQueryBuilder.java

```diff
@@ -0,0 +1,263 @@
+        if (query == null) {
+            throw new IllegalArgumentException("[" + NAME + "] requires a " + QUERY_FIELD.getPreferredName() + " value");
+        }
+        this.fieldName = fieldName;
+        this.query = query;
+    }
+
+    public SemanticQueryBuilder(StreamInput in) throws IOException {
+        super(in);
+        this.fieldName = in.readString();
+        this.query = in.readString();
+        this.inferenceResults = in.readOptionalNamedWriteable(InferenceServiceResults.class);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        if (inferenceResultsSupplier != null) {
+            throw new IllegalStateException("Inference results supplier is set. Missing a rewriteAndFetch?");
+        }
+        out.writeString(fieldName);
+        out.writeString(query);
+        out.writeOptionalNamedWriteable(inferenceResults);
+    }
+
+    private SemanticQueryBuilder(
+        SemanticQueryBuilder other,
+        SetOnce<InferenceServiceResults> inferenceResultsSupplier,
+        InferenceServiceResults inferenceResults
+    ) {
+        this.fieldName = other.fieldName;
+        this.query = other.query;
+        this.boost = other.boost;
+        this.queryName = other.queryName;
+        this.inferenceResultsSupplier = inferenceResultsSupplier;
+        this.inferenceResults = inferenceResults;
+    }
+
+    @Override
+    public String getWriteableName() {
+        return NAME;
+    }
+
+    @Override
+    public TransportVersion getMinimalSupportedVersion() {
+        return TransportVersions.SEMANTIC_QUERY;
+    }
+
+    public static SemanticQueryBuilder fromXContent(XContentParser parser) throws IOException {
+        return PARSER.apply(parser, null);
+    }
+
+    @Override
+    protected void doXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject(NAME);
+        builder.field(FIELD_FIELD.getPreferredName(), fieldName);
+        builder.field(QUERY_FIELD.getPreferredName(), query);
+        boostAndQueryNameToXContent(builder);
+        builder.endObject();
+    }
+
+    @Override
+    protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) {
+        SearchExecutionContext searchExecutionContext = queryRewriteContext.convertToSearchExecutionContext();
+        if (searchExecutionContext != null) {
+            return doRewriteBuildSemanticQuery(searchExecutionContext);
+        }
+
+        return doRewriteGetInferenceResults(queryRewriteContext);
+    }
+
+    private QueryBuilder doRewriteBuildSemanticQuery(SearchExecutionContext searchExecutionContext) {
+        if (inferenceResults == null) {
+            throw new IllegalStateException("Query builder must have inference results before rewriting to another query type");
+        }
+
+        MappedFieldType fieldType = searchExecutionContext.getFieldType(fieldName);
+        if (fieldType == null) {
+            return new MatchNoneQueryBuilder();
```

Comment:
What is the reason we don't throw `IllegalArgumentException` in this case?

Reply:
We return `MatchNoneQueryBuilder` here so that the query respects the [index.query.parse.allow_unmapped_fields](https://github.com/elastic/elasticsearch/blob/b6b20a5d6f3580b13803ca5c660584ded829d6a8/server/src/main/java/org/elasticsearch/index/IndexSettings.java#L82) setting. If this is `true` (the default), queries should return no results when referring to an unmapped field. If this is `false`, `searchExecutionContext.getFieldType` will throw an exception on an unmapped field.

Reply:
Thanks Mike,  we don't document this setting and don't consistently use it across queries.  My question is about your opinion if for `semantic` query it makes sense to return 0 docs if field is not mapped?  

To my mind, it makes more sense to return an error in case a user misspelled a field name and expects results. WDYT?

Reply:
I think we should respect the setting, even though it's undocumented. We can & should bring it up for a larger discussion, but the fact that https://github.com/elastic/elasticsearch/issues/106846 was accepted as a legitimate bug (with a fix in progress) is a strong hint to me that we should respect the setting.

Reply:
+1 for bringing up for a larger discussion (added to the agenda).